File: frontend/src/api/assignments.js
import api from "./axios";

export const listForLesson = (lessonId) =>
  api.get(`/api/assignments/lesson/${lessonId}`).then((res) => res.data);

export const createAssignment = (payload) => api.post("/api/assignments", payload).then((res) => res.data);

export const submitAssignment = (assignmentId, file) => {
  const formData = new FormData();
  formData.append("file", file);
  return api
    .post(`/api/assignments/${assignmentId}/submit`, formData, {
      headers: { "Content-Type": "multipart/form-data" },
    })
    .then((res) => res.data);
};

export const listSubmissions = (assignmentId) =>
  api.get(`/api/assignments/${assignmentId}/submissions`).then((res) => res.data);

export const mySubmissions = () => api.get("/api/assignments/submissions/me").then((res) => res.data);

export const gradeSubmission = (submissionId, grade, feedback) =>
  api.patch(`/api/assignments/submissions/${submissionId}/grade`, { grade, feedback }).then((res) => res.data);

File: backend/app/routers/assignments.py
from fastapi import APIRouter, Depends, File, UploadFile
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.core.deps import require_roles
from app.models.user import User, UserRole
from app.schemas.assignment import AssignmentCreate, AssignmentOut, GradeSubmissionRequest, SubmissionOut
from app.services import assignment_service

router = APIRouter(prefix="/api/assignments", tags=["assignments"])

staff_only = require_roles(UserRole.HEAD_ADMIN, UserRole.MODERATOR, UserRole.INSTRUCTOR)
student_only = require_roles(UserRole.STUDENT)


@router.get("/lesson/{lesson_id}", response_model=list[AssignmentOut])
def list_for_lesson(lesson_id: int, db: Session = Depends(get_db)):
    return assignment_service.list_assignments_for_lesson(db, lesson_id)


@router.post("", response_model=AssignmentOut, status_code=201)
def create_assignment(data: AssignmentCreate, db: Session = Depends(get_db), current_user: User = Depends(staff_only)):
    return assignment_service.create_assignment(db, creator_id=current_user.id, data=data)


@router.post("/{assignment_id}/submit", response_model=SubmissionOut, status_code=201)
def submit(
    assignment_id: int,
    file: UploadFile = File(...),
    db: Session = Depends(get_db),
    current_user: User = Depends(student_only),
):
    return assignment_service.submit_assignment(db, assignment_id, current_user.id, file)


@router.get("/{assignment_id}/submissions", response_model=list[SubmissionOut], dependencies=[Depends(staff_only)])
def list_submissions(assignment_id: int, db: Session = Depends(get_db)):
    return assignment_service.list_submissions(db, assignment_id)


@router.get("/submissions/me", response_model=list[SubmissionOut])
def my_submissions(db: Session = Depends(get_db), current_user: User = Depends(student_only)):
    return assignment_service.my_submissions(db, current_user.id)


@router.patch("/submissions/{submission_id}/grade", response_model=SubmissionOut)
def grade(
    submission_id: int,
    data: GradeSubmissionRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(staff_only),
):
    return assignment_service.grade_submission(db, submission_id, current_user.id, data)
